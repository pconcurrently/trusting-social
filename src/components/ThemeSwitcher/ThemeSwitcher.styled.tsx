import styled, { keyframes } from 'styled-components';

const spin = keyframes`
  from { transform: rotate(0deg); }
  to { transform: rotate(360deg); }
`;

const shake = keyframes`
  0% { transform: translate(1px, 1px) rotate(0deg); }
  50% { transform: translate(1px, -2px) rotate(1deg); }
  100% { transform: translate(1px, 1px) rotate(0deg); }
`;

export const StyledThemeSwitcher = styled.div`
  display: flex;
  align-items: center;

  .theme-switch {
    display: inline-block;
    height: 28px;
    position: relative;
    width: 60px;
  }

  .theme-switch input {
    display: none;
  }

  .slider {
    background-color: #555555;
    bottom: 0;
    cursor: pointer;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
    transition: 0.4s;
  }

  .slider:before {
    background-color: #fff;
    bottom: 4px;
    content: '';
    height: 20px;
    left: 4px;
    position: absolute;
    transition: 0.4s;
    width: 20px;
  }

  img {
    height: 20px;
    position: absolute;
    top: 4px;

    &.light-icon {
      right: 4px;
      opacity: 1;
      transition: 0.4s;
      animation: ${spin} 4s linear infinite;
    }

    &.dark-icon {
      left: 4px;
      opacity: 0;
      transition: 0.4s;
      animation: ${shake} 2s linear infinite;
      fill: #FFF069;
    }
  }

  input:checked + .slider {
    background-color: #555555;
    img.light-icon {
      opacity: 0;
    }
    img.dark-icon {
      opacity: 1;
    }
  }

  input:checked + .slider:before {
    transform: translateX(32px);
  }

  .slider.round {
    border-radius: 34px;
  }

  .slider.round:before {
    border-radius: 50%;
  }
`;
