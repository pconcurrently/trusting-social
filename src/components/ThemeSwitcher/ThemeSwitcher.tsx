import React from 'react';

import { StyledThemeSwitcher } from './ThemeSwitcher.styled';
import sunnyIcon from '../../images/sunny-icon.svg';
import nightIcon from '../../images/night-icon.svg';

interface IProps {
  onClick: () => void;
}

export const ThemeSwitcher = (props: IProps) => {
  const { onClick } = props;
  return (
    <StyledThemeSwitcher>
      <label className="theme-switch" htmlFor="checkbox">
        <input type="checkbox" id="checkbox" onChange={onClick}/>
        <div className="slider round">
          <img className="light-icon" src={sunnyIcon} alt="light"/>
          <img className="dark-icon" src={nightIcon} alt="dark"/>
        </div>
      </label>
    </StyledThemeSwitcher>
  );
};
