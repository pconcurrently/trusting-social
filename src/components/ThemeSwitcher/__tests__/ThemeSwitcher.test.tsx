import React from 'react';
import { mount, shallow } from 'enzyme';
import { ThemeSwitcher } from '../ThemeSwitcher';

describe('App', () => {
  const onClick = jest.fn();

  it('should render ThemeSwitcher', () => {
    const container = shallow(<ThemeSwitcher onClick={onClick} />);
    expect(container).toMatchSnapshot();
  });

  it('should fire onClick function', () => {
    const container = mount(<ThemeSwitcher onClick={onClick} />);
    const event = { currentTarget: { checked: true } };
    expect(container.find('input#checkbox').props().checked).toBeFalsy();
    container.find('input#checkbox').simulate('change', event);
    expect(onClick).toBeCalled();
  });
});
