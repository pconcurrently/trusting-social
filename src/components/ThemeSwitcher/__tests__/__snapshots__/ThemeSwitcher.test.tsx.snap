// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render ThemeSwitcher 1`] = `
"<styled.div>
  <label className=\\"theme-switch\\" htmlFor=\\"checkbox\\">
    <input type=\\"checkbox\\" id=\\"checkbox\\" onChange={[Function: mockConstructor]} />
    <div className=\\"slider round\\">
      <img className=\\"light-icon\\" src=\\"sunny-icon.svg\\" alt=\\"light\\" />
      <img className=\\"dark-icon\\" src=\\"night-icon.svg\\" alt=\\"dark\\" />
    </div>
  </label>
</styled.div>"
`;
